File: wisefy/src/androidTest/java/com/isupatches/wisefy/test/GetFrequencyTests.java
package com.isupatches.wisefy.test;


import android.net.wifi.WifiInfo;
import android.os.Build;
import com.isupatches.wisefy.callbacks.GetFrequencyCallbacks;
import com.isupatches.wisefy.constants.WiseFyCodes;
import com.isupatches.wisefy.test.base.BaseTestClass;
import com.isupatches.wisefy.test.base.TestActivity;
import org.junit.Test;
import static com.isupatches.wisefy.test.base.TestUtils.TEST_NETWORK_FREQUENCY_24GHZ;
import static junit.framework.Assert.assertEquals;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.timeout;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;


public class GetFrequencyTests extends BaseTestClass<TestActivity> {

    public GetFrequencyTests() {
        super(TestActivity.class);
    }

    @Test
    public void noCallbacks_getFrequency_currentNetwork_failure() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(null);
            assertEquals(null, mWiseFy.getFrequency());
        }
    }

    @Test
    public void noCallbacks_getFrequency_currentNetwork_success() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            WifiInfo mockWifiInfo = mock(WifiInfo.class);
            when(mockWifiInfo.getFrequency()).thenReturn(TEST_NETWORK_FREQUENCY_24GHZ);
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(mockWifiInfo);
            assertEquals(TEST_NETWORK_FREQUENCY_24GHZ, (int) mWiseFy.getFrequency());
        }
    }

    @Test
    public void callbacks_getFrequency_currentNetwork_failure() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(null);
            GetFrequencyCallbacks mockGetFrequencyCallbacks = mock(GetFrequencyCallbacks.class);
            mWiseFy.getFrequency(mockGetFrequencyCallbacks);
            verify(mockGetFrequencyCallbacks, timeout(2000)).failureGettingFrequency();
        }
    }

    @Test
    public void callbacks_getFrequency_currentNetwork_failure_nullCallback() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(null);
            GetFrequencyCallbacks mockGetFrequencyCallbacks = mock(GetFrequencyCallbacks.class);
            mWiseFy.getFrequency((GetFrequencyCallbacks) null);
            verify(mockGetFrequencyCallbacks, never()).failureGettingFrequency();
        }
    }

    @Test
    public void callbacks_getFrequency_currentNetwork_success() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            WifiInfo mockWifiInfo = mock(WifiInfo.class);
            when(mockWifiInfo.getFrequency()).thenReturn(TEST_NETWORK_FREQUENCY_24GHZ);
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(mockWifiInfo);
            GetFrequencyCallbacks mockGetFrequencyCallbacks = mock(GetFrequencyCallbacks.class);
            mWiseFy.getFrequency(mockGetFrequencyCallbacks);
            verify(mockGetFrequencyCallbacks, timeout(2000)).retrievedFrequency(TEST_NETWORK_FREQUENCY_24GHZ);
        }
    }

    @Test
    public void callbacks_getFrequency_currentNetwork_success_nullCallback() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            WifiInfo mockWifiInfo = mock(WifiInfo.class);
            when(mockWifiInfo.getFrequency()).thenReturn(TEST_NETWORK_FREQUENCY_24GHZ);
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(mockWifiInfo);
            GetFrequencyCallbacks mockGetFrequencyCallbacks = mock(GetFrequencyCallbacks.class);
            mWiseFy.getFrequency((GetFrequencyCallbacks) null);
            verify(mockGetFrequencyCallbacks, never()).retrievedFrequency(TEST_NETWORK_FREQUENCY_24GHZ);
        }
    }

    @Test
    public void noCallbacks_getFrequency_provideWifiInfo_failure() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(null);
            assertEquals(null, mWiseFy.getFrequency((WifiInfo) null));
        }
    }

    @Test
    public void noCallbacks_getFrequency_provideWifiInfo_success() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            WifiInfo mockWifiInfo = mock(WifiInfo.class);
            when(mockWifiInfo.getFrequency()).thenReturn(TEST_NETWORK_FREQUENCY_24GHZ);
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(mockWifiInfo);
            assertEquals(TEST_NETWORK_FREQUENCY_24GHZ, (int) mWiseFy.getFrequency(mockWifiInfo));
        }
    }

    @Test
    public void callbacks_getFrequency_provideWifiInfo_failure() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(null);
            GetFrequencyCallbacks mockGetFrequencyCallbacks = mock(GetFrequencyCallbacks.class);
            mWiseFy.getFrequency(null, mockGetFrequencyCallbacks);
            verify(mockGetFrequencyCallbacks, timeout(2000)).getFrequencyWiseFyFailure(WiseFyCodes.MISSING_PARAMETER);
        }
    }

    @Test
    public void callbacks_getFrequency_provideWifiInfo_failure_nullCallback() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(null);
            GetFrequencyCallbacks mockGetFrequencyCallbacks = mock(GetFrequencyCallbacks.class);
            mWiseFy.getFrequency(null, null);
            verify(mockGetFrequencyCallbacks, never()).failureGettingFrequency();
        }
    }

    @Test
    public void callbacks_getFrequency_provideWifiInfo_success() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            WifiInfo mockWifiInfo = mock(WifiInfo.class);
            when(mockWifiInfo.getFrequency()).thenReturn(TEST_NETWORK_FREQUENCY_24GHZ);
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(mockWifiInfo);
            GetFrequencyCallbacks mockGetFrequencyCallbacks = mock(GetFrequencyCallbacks.class);
            mWiseFy.getFrequency(mockWifiInfo, mockGetFrequencyCallbacks);
            verify(mockGetFrequencyCallbacks, timeout(2000)).retrievedFrequency(TEST_NETWORK_FREQUENCY_24GHZ);
        }
    }

    @Test
    public void callbacks_getFrequency_provideWifiInfo_success_nullCallback() {
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            setManagers();
            WifiInfo mockWifiInfo = mock(WifiInfo.class);
            when(mockWifiInfo.getFrequency()).thenReturn(TEST_NETWORK_FREQUENCY_24GHZ);
            when(mMockWiFiManager.getConnectionInfo()).thenReturn(mockWifiInfo);
            GetFrequencyCallbacks mockGetFrequencyCallbacks = mock(GetFrequencyCallbacks.class);
            mWiseFy.getFrequency(mockWifiInfo, null);
            verify(mockGetFrequencyCallbacks, never()).retrievedFrequency(TEST_NETWORK_FREQUENCY_24GHZ);
        }
    }
}

File: wisefy/src/test/java/com/isupatches/wisefy/GetFrequencyTests.kt
package com.isupatches.wisefy

import android.net.wifi.WifiInfo
import android.os.Build

import com.isupatches.wisefy.callbacks.GetFrequencyCallbacks
import com.isupatches.wisefy.constants.MISSING_PARAMETER
import com.isupatches.wisefy.internal.base.BaseUnitTest

import org.junit.Assert.assertEquals
import org.junit.Assert.fail
import org.junit.Test
import org.junit.runner.RunWith

import org.mockito.Mockito.after
import org.mockito.Mockito.mock
import org.mockito.Mockito.timeout
import org.mockito.Mockito.verify

import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

/**
 * Tests the ability to retrieve a network's frequency.
 *
 * @author Patches
 */
@RunWith(RobolectricTestRunner::class)
@Config(sdk = [Build.VERSION_CODES.LOLLIPOP])
internal class GetFrequencyTests : BaseUnitTest() {

    @Test fun sync_getFrequency_currentNetwork_failure() {
        mockWiseFyPrechecksUtil.getCurrentNetwork_success()
        mockNetworkUtil.currentNetwork_null()
        assertEquals(null, wisefy.getFrequency())
        verificationUtil.triedToGetCurrentNetwork()
    }

    @Test fun sync_getFrequency_currentNetwork_success() {
        mockWiseFyPrechecksUtil.getCurrentNetwork_success()
        val wifiInfo = mockNetworkUtil.networkWithFrequency(TEST_NETWORK_FREQUENCY_24GHZ)
        val frequency = wisefy.getFrequency()
        if (frequency != null) {
            assertEquals(TEST_NETWORK_FREQUENCY_24GHZ.toLong(), frequency.toLong())
            verificationUtil.triedToGetCurrentNetwork()
            verify(wifiInfo, timeout(VERIFICATION_SUCCESS_TIMEOUT)).frequency
        } else {
            fail()
        }
    }

    @Test fun async_getFrequency_currentNetwork_failure() {
        mockWiseFyPrechecksUtil.getCurrentNetwork_success()
        mockNetworkUtil.currentNetwork_null()
        val mockCallbacks = mock(GetFrequencyCallbacks::class.java)
        wisefy.getFrequency(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).failureGettingFrequency()
        verificationUtil.triedToGetCurrentNetwork()
    }

    @Test fun async_getFrequency_currentNetwork_failure_nullCallback() {
        mockWiseFyPrechecksUtil.getCurrentNetwork_success()
        mockNetworkUtil.currentNetwork_null()
        nullCallbackUtil.callGetFrequency()
        verificationUtil.triedToGetCurrentNetwork()
    }

    @Test fun async_getFrequency_currentNetwork_success() {
        mockWiseFyPrechecksUtil.getCurrentNetwork_success()
        val mockWifiInfo = mockNetworkUtil.networkWithFrequency(TEST_NETWORK_FREQUENCY_24GHZ)
        val mockCallbacks = mock(GetFrequencyCallbacks::class.java)
        wisefy.getFrequency(mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedFrequency(TEST_NETWORK_FREQUENCY_24GHZ)
        verificationUtil.triedToGetCurrentNetwork()
        verify(mockWifiInfo, timeout(VERIFICATION_SUCCESS_TIMEOUT)).frequency
    }

    @Test fun async_getFrequency_currentNetwork_success_nullCallback() {
        mockWiseFyPrechecksUtil.getCurrentNetwork_success()
        mockNetworkUtil.networkWithFrequency(TEST_NETWORK_FREQUENCY_24GHZ)
        nullCallbackUtil.callGetFrequency()
        verificationUtil.triedToGetCurrentNetwork()
    }

    @Test fun sync_getFrequency_networkProvided_failure() {
        mockNetworkUtil.currentNetwork_null()
        assertEquals(null, wisefy.getFrequency(null as? WifiInfo?))
    }

    @Test fun sync_getFrequency_networkProvided_success() {
        val mockWifiInfo = mockNetworkUtil.networkWithFrequency(TEST_NETWORK_FREQUENCY_24GHZ)
        val frequency = wisefy.getFrequency(mockWifiInfo)
        if (frequency != null) {
            assertEquals(TEST_NETWORK_FREQUENCY_24GHZ.toLong(), frequency.toLong())
        } else {
            fail()
        }
    }

    @Test fun async_getFrequency_networkProvided_failure() {
        mockNetworkUtil.currentNetwork_null()
        val mockCallbacks = mock(GetFrequencyCallbacks::class.java)
        wisefy.getFrequency(null, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).wisefyFailure(MISSING_PARAMETER)
    }

    @Test fun async_getFrequency_networkProvided_failure_nullCallback() {
        mockNetworkUtil.currentNetwork_null()
        nullCallbackUtil.callGetFrequency_networkProvided(null)
    }

    @Test fun async_getFrequency_networkProvided_success() {
        val mockWifiInfo = mockNetworkUtil.networkWithFrequency(TEST_NETWORK_FREQUENCY_24GHZ)
        val mockCallbacks = mock(GetFrequencyCallbacks::class.java)
        wisefy.getFrequency(mockWifiInfo, mockCallbacks)
        verify(mockCallbacks, timeout(VERIFICATION_SUCCESS_TIMEOUT)).retrievedFrequency(TEST_NETWORK_FREQUENCY_24GHZ)
        verify(mockWifiInfo, timeout(VERIFICATION_SUCCESS_TIMEOUT)).frequency
    }

    @Test fun async_getFrequency_networkProvided_success_nullCallback() {
        val mockWifiInfo = mockNetworkUtil.networkWithFrequency(TEST_NETWORK_FREQUENCY_24GHZ)
        nullCallbackUtil.callGetFrequency_networkProvided(mockWifiInfo)
        verify(mockWifiInfo, after(VERIFICATION_FAILURE_TIMEOUT).times(0)).frequency
    }
}
